File: app/src/main/java/com/treelzebub/umap/util/UMapBus.java
package com.treelzebub.umap.util;

import android.os.Handler;
import android.os.Looper;

import com.squareup.otto.Bus;

/**
 * Created by Tre Murillo on 5/16/15
 * Copyright(c) 2015 Level, Inc.
 */
public class UMapBus extends Bus {
    private final Handler mainThread = new Handler(Looper.getMainLooper());

    @Override
    public void post(final Object event) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            super.post(event);
        } else {
            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    post(event);
                }
            });
        }
    }

    @Override
    public void register(final Object object) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            super.register(object);
        } else {
            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    register(object);
                }
            });
        }
    }

    @Override
    public void unregister(final Object object) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            try {
                super.unregister(object);
            } catch (IllegalArgumentException e) {
                // dont care
            }
        } else {
            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    unregister(object);
                }
            });
        }
    }
}

File: app/src/main/kotlin/com/treelzebub/umap/util/UMapBus.kt
package com.treelzebub.umap.util

import android.os.Handler
import android.os.Looper
import com.squareup.otto.Bus

/**
 * Created by Tre Murillo on 5/28/15
 * Copyright(c) 2015 Level, Inc.
 */
public class UMapBus : Bus() {
private val mainThread = Handler(Looper.getMainLooper())

override fun post(event: Any) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
        super.post(event)
    } else {
        mainThread.post(object : Runnable {
            override fun run() {
                post(event)
            }
        })
    }
}

override fun register(any: Any) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
        super.register(any)
    } else {
        mainThread.post(object : Runnable {
            override fun run() {
                register(any)
            }
        })
    }
}

override fun unregister(any: Any) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
        try {
            super.unregister(any)
        } catch (e: IllegalArgumentException) {
            // dont care
        }

    } else {
        mainThread.post(object : Runnable {
            override fun run() {
                unregister(any)
            }
        })
    }
}
}
